### PR TITLE
add lmod_update_system_cache.sh script in contrib

### PR DESCRIPTION
**Pull request by boegel**
This script is a rewrite of the existing createSystemCacheFile.sh. I feel this is better organised, and since it allows for verbose debug logging to stdout, it's easier to use.

It now hardcodes a couple of settings, that should be modified. I'll indicate what should be changed by making inline remarks.

### CODE REVIEW

File: contrib/BuildSystemCacheFile/lmod_update_system_cache.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    # include Lua version in filename of compiled cache, to avoid compatibility issues when Lua is upgraded
+    lua_ver=$(lua -e 'print((_VERSION:gsub("Lua ","")))')
+    luac -o ${cache_file_name}.new.luac_$lua_ver ${cache_file_name}.lua
+    install_new_cache luac_$lua_ver ${cache_file_name}
+}
+
+
+# figure out location of Lmod cache and cache timestamp file via 'ml --config'
+LMOD_CACHE_DIR=`ml --config 2>&1 | grep -A 2 "Cache Directory" | tail -1 | tr -s ' ' | cut -f1 -d' '`
+LMOD_CACHE_TIMESTAMP_FILE=`ml --config 2>&1 | grep -A 2 "Cache Directory" | tail -1 | tr -s ' ' | cut -f2 -d' '`
```
Comment (boegel) [inline contrib/BuildSystemCacheFile/lmod_update_system_cache.sh]: these are good defaults imho, but it should be possible to tweak these

so, these should become `LMOD_CACHE_DIR_DEFAULT` and `LMOD_CACHE_TIMESTAMP_FILE_DEFAULT`, and then `LMOD_CACHE_DIR` and `LMOD_CACHE_TIMESTAMP_FILE` should be picking up those default values unless they're already defined

File: contrib/BuildSystemCacheFile/lmod_update_system_cache.sh [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+mkdir -p `dirname $LMOD_CACHE_TIMESTAMP_FILE`
+
+# generate new cache timestamp file
+# timestamp file marks time of last change to the system, and *must* be older than the Lmod cache itself
+debug "Cleaning up ${LMOD_CACHE_TIMESTAMP_FILE}.*"
+debug "`ls -l ${LMOD_CACHE_TIMESTAMP_FILE}.* 2> /dev/null`"
+rm -rf ${LMOD_CACHE_TIMESTAMP_FILE}.new
+
+debug "Updating Lmod cache timestamp file @ ${LMOD_CACHE_TIMESTAMP_FILE}.new"
+touch ${LMOD_CACHE_TIMESTAMP_FILE}.new  # leave timestamp file empty on purpose, because Lmod uses contents to try to do magic
```
Comment (boegel) [inline contrib/BuildSystemCacheFile/lmod_update_system_cache.sh]: add support for using a non-empty timestamp file?

File: contrib/BuildSystemCacheFile/lmod_update_system_cache.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+update_cache moduleT
+update_cache dbT
```
Comment (boegel) [inline contrib/BuildSystemCacheFile/lmod_update_system_cache.sh]: this should be optional (steered by `LMOD_BUILD_DBT_CACHE`?), probably disabled by default

File: contrib/BuildSystemCacheFile/lmod_update_system_cache.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+#update_cache reverseMapT
```
Comment (boegel) [inline contrib/BuildSystemCacheFile/lmod_update_system_cache.sh]: this should be steered by something like `LMOD_BUILD_REVERSEMAPT_CACHE`

**Comment by boegel** [comment]
@rtmclay: thoughts on this?

**Comment by rtmclay** [comment]
This script need to take arguments to specify the following:

   * The cache directory
   * The time stamp file
   * The part of  MODULEPATH that goes in the cache file
   * A debug flag 
   * What to build  moduleT, dbT,  reverseMapT?

I think that it should build moduleT, dbT by default.  But it should be able to build the reverseMapT if asked.  @boegel: comments?

**Comment by rtmclay** [comment]
Also there should be an option which says extract the locations from module --config.  It should only work when there is only one cache directory and timestamp file.

**Comment by boegel** [comment]
@rtmclay: I've pushed a new version of the script. There are a couple of minor things I would like to see changed before this is merged in, see the `FIXME`s in the script.

Any suggestions for other improvements are welcome, but I like where this is going.

**Comment by boegel** [comment]
@rtmclay: we'll also need to update the `README` in that directory, the one that's in there now is outdated anyway (the trick with the non-empty timestamp file no longer works, as you informed me)

**Pull request merged**